Sources/TABResourceLoader/Services/NetworkResponseHandler.swift

```diff
@@ -38,51 +37,56 @@ import Foundation
 ///   5. Could not parse the data
 public enum NetworkResponse<Model> {
   case success(Model, HTTPURLResponse)
-  case failure(Result<Model>, HTTPURLResponse?, NetworkServiceError)
-}
-
-enum NetworkResponseHandlerError: Error {
-  case noDataProvided
+  case failure(NetworkServiceError, HTTPURLResponse?)
 }
 
 struct NetworkResponseHandler {
 
   static func resultFrom<Resource: NetworkResourceType & DataResourceType>(resource: Resource, data: Data?, URLResponse: Foundation.URLResponse?, error: Error?) -> NetworkResponse<Resource.Model> {
 
-    let parsedResult = self.parsedResult(transformedFrom: data, using: resource)
-
     guard let HTTPURLResponse = URLResponse as? HTTPURLResponse else {
       if let error = error {
-        return .failure(parsedResult, nil, .sessionError(error: error))
+        return .failure(.sessionError(error: error), nil)
       } else {
-        return .failure(parsedResult, nil, .noHTTPURLResponse)
+        return .failure(.noHTTPURLResponse, nil)
       }
     }
 
-    switch HTTPURLResponse.statusCode {
-    case 400..<600:
-      return .failure(parsedResult, HTTPURLResponse, .statusCodeError(statusCode: HTTPURLResponse.statusCode))
-    default: break
+    guard let data = data else {
+      let networkError = self.networkServiceError(HTTPURLResponse: HTTPURLResponse, error: error) ?? NetworkServiceError.noDataProvided
+      return .failure(networkError, HTTPURLResponse)
     }
 
-    if let error = error {
-      return .failure(parsedResult, HTTPURLResponse, .sessionError(error: error))
+    switch HTTPURLResponse.statusCode {
+    case 200..<400:
+      let parsedResult = Result<Resource.Model> {
+        return try resource.model(from: data)
+      }
+      switch parsedResult {
+      case .success(let model):
+        return .success(model, HTTPURLResponse)
+      case .failure:
+        return .failure(.statusCodeError(statusCode: HTTPURLResponse.statusCode), HTTPURLResponse)
+      }
+    default:
+      if let error = resource.error(from: data) {
+        return .failure(.couldNotParseModel(error: error), HTTPURLResponse)
```

Comment:
I think probably `statusCodeCustomError` is better since we don't want to an optional `Error` as that will always exist but we can have a `statusCodeError` without a custom error.